tools/wasi/wasi.cpp

```diff
@@ -0,0 +1,235 @@
+{
+    uint32_t environc = static_cast<uint32_t>(args[0]);
+    // TODO: implement properly (only returns 0 now)
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), environc, 0);
+    return {false, {UVWASI_ESUCCESS}};
+}
+
+// TODO: support both wasi_snapshot_preview1 and wasi_unstable
+const std::vector<fizzy::ImportedFunction> wasi_functions = {
+    {"wasi_snapshot_preview1", "proc_exit", {fizzy::ValType::i32}, std::nullopt, wasi_proc_exit},
+    // clock_time_get
+    // fd_read
+    {"wasi_snapshot_preview1", "fd_read",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_read},
+    {"wasi_snapshot_preview1", "fd_write",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_write},
+    {"wasi_snapshot_preview1", "fd_prestat_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_prestat_get},
+    {"wasi_snapshot_preview1", "fd_prestat_dir_name",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32}, fizzy::ValType::i32,
+        wasi_return_enosys},
+    {"wasi_snapshot_preview1", "environ_sizes_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_environ_sizes_get},
+    {"wasi_snapshot_preview1", "environ_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_return_enosys},
+
+
+    {"wasi_unstable", "proc_exit", {fizzy::ValType::i32}, std::nullopt, wasi_proc_exit},
+    // clock_time_get
+    // fd_read
+    {"wasi_unstable", "clock_time_get",
+        {fizzy::ValType::i32, fizzy::ValType::i64, fizzy::ValType::i32}, fizzy::ValType::i32,
+        wasi_return_enosys},
+    {"wasi_unstable", "fd_read",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_read},
+    {"wasi_unstable", "fd_write",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_write},
+    {"wasi_unstable", "fd_prestat_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_prestat_get},
+    {"wasi_unstable", "fd_prestat_dir_name",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32}, fizzy::ValType::i32,
+        wasi_return_enosys},
+    {"wasi_unstable", "environ_sizes_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_environ_sizes_get},
+    {"wasi_unstable", "environ_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_return_enosys},
+};
+
+bool run(int argc, char** argv)
+{
+    uvwasi_options_t options = {
+        3,           // sizeof fd_table
+        0, nullptr,  // NOTE: no remappings
+        static_cast<uvwasi_size_t>(argc), const_cast<const char**>(argv),
```

Comment:
does it work without `const_cast`?
```suggestion
        static_cast<uvwasi_size_t>(argc), argv,
```

Reply:
For some reason this doesn't work for me unless I change the function signature of `run`.

Reply:
@gumb0 any comment?

Reply:
Yeah, looks like there's no way around cast here

https://isocpp.org/wiki/faq/const-correctness#constptrptr-conversion
http://c-faq.com/ansi/constmismatch.html

Reply:
Did you try declaring `main()` as `int main(int c, const char**v)`?

Reply:
Sure, I can do that. In a previous PR @gumb0 mentioned that `main` cannot be declared noexcept and perhaps const could pose such a problem too: https://en.cppreference.com/w/cpp/language/main_function

But changing both run and main to const avoids the cast.

Reply:
I'm fine with `char**`.